src/ElfReader.cc

```diff
@@ -127,16 +161,85 @@ const typename Arch::ElfShdr* ElfReaderImpl<Arch>::find_section(const char* n) {
 template <typename Arch>
 SectionOffsets ElfReaderImpl<Arch>::find_section_file_offsets(
     const char* name) {
-  SectionOffsets offsets = { 0, 0 };
+  SectionOffsets offsets = { 0, 0, false };
   const typename Arch::ElfShdr* section = find_section(name);
   if (!section) {
     return offsets;
   }
   offsets.start = section->sh_offset;
   offsets.end = section->sh_offset + section->sh_size;
+  offsets.compressed = !!(section->sh_flags & SHF_COMPRESSED);
   return offsets;
 }
 
+template <typename Arch>
+DecompressedSection* ElfReaderImpl<Arch>::decompress_section(SectionOffsets offsets) {
+  DEBUG_ASSERT(offsets.compressed);
+  auto start = offsets.start;
+  auto hdr = r.read<typename Arch::ElfChdr>(offsets.start);
+  if (!hdr) {
+    LOG(warn) << "section at " << offsets.start
+              << " is marked compressed but is too small";
+    return nullptr;
+  }
+
+  size_t decompressed_size = 0;
+  if (hdr->ch_type == ELFCOMPRESS_ZLIB) {
+    decompressed_size = hdr->ch_size;
+    offsets.start += sizeof(typename Arch::ElfChdr);
+  } else {
+    auto legacy_hdr = r.read_bytes(offsets.start, 4);
+    if (!memcmp("ZLIB", legacy_hdr, 4)) {
+      auto be_size = r.read<uint64_t>(offsets.start + 4);
+      decompressed_size = be64toh(*be_size);
+      offsets.start += 12;
+    } else {
+      LOG(warn) << "section at " << offsets.start
+                << " is marked compressed but uses unrecognized"
+                << " type " << HEX(hdr->ch_type);
+      return nullptr;
+    }
+  }
+
+  uint8_t* buf = (uint8_t*)mmap(NULL, decompressed_size, PROT_READ | PROT_WRITE,
+                                MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+  if (!buf) {
+    FATAL() << "OOM";
+    return nullptr;
+  }
+
+
+  z_stream stream;
+  memset(&stream, 0, sizeof(stream));
+  int result = inflateInit(&stream);
+  if (result != Z_OK) {
+    FATAL() << "inflateInit failed!";
+    return nullptr;
+  }
+
+  stream.avail_in = offsets.end - offsets.start;
+  stream.next_in = (unsigned char*)r.read_bytes(offsets.start, stream.avail_in);
+  stream.next_out = buf;
+  stream.avail_out = decompressed_size;
+  result = inflate(&stream, Z_FINISH);
+  if (result != Z_STREAM_END) {
+    FATAL() << "inflate failed!";
+    return nullptr;
+  }
+
+  result = inflateEnd(&stream);
+  if (result != Z_OK) {
+    FATAL() << "inflateEnd failed!";
+    return nullptr;
+  }
+
+  mprotect(buf, decompressed_size, PROT_READ);
```

Comment:
Hmm, so this is why you're using an mmap region? Why is this `mprotect` important?

Reply:
I made it PROT_READ to match the "normal" mmapped-in binary from ElfFileReader. If we decide we don't care about this we could just use vector<uint8_t>.

Reply:
I think we don't care